ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    public void save(Message message, String parentId, Integer traceNum, String action, ActionListener listener) {
+        ConversationIndexMessage msg = (ConversationIndexMessage) message;
+        memoryManager
+            .createInteraction(conversationId, msg.getQuestion(), null, msg.getResponse(), action, null, parentId, traceNum, listener);
+    }
+
+    @Override
+    public void getMessages(String id, ActionListener listener) {
+        SearchRequest searchRequest = new SearchRequest();
+        searchRequest.indices(memoryMessageIndexName);
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        sourceBuilder.size(10000);
+        QueryBuilder sessionIdQueryBuilder = new TermQueryBuilder(CONVERSATION_ID, id);
+
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.must(sessionIdQueryBuilder);
+
+        if (retrieveFinalAnswer) {
+            QueryBuilder finalAnswerQueryBuilder = new TermQueryBuilder(FINAL_ANSWER, true);
+            boolQueryBuilder.must(finalAnswerQueryBuilder);
+        }
+
+        sourceBuilder.query(boolQueryBuilder);
+        sourceBuilder.sort(CREATED_TIME, SortOrder.ASC);
+        searchRequest.source(sourceBuilder);
+        client.search(searchRequest, listener);
+    }
+
+    public void getMessages(ActionListener listener) {
+        memoryManager.getFinalInteractions(conversationId, LAST_N_INTERACTIONS, listener);
+    }
+
+    @Override
+    public void clear() {
+        throw new RuntimeException("clear method is not supported in ConversationIndexMemory");
+    }
+
+    @Override
+    public void remove(String id) {
+        throw new RuntimeException("remove method is not supported in ConversationIndexMemory");
+    }
+
+    public static class Factory implements Memory.Factory<ConversationIndexMemory> {
+        private Client client;
+        private MLIndicesHandler mlIndicesHandler;
+        private String memoryMetaIndexName = ML_MEMORY_META_INDEX;
+        private String memoryMessageIndexName = ML_MEMORY_MESSAGE_INDEX;
+        private MLMemoryManager memoryManager;
+
+        public void init(Client client, MLIndicesHandler mlIndicesHandler, MLMemoryManager memoryManager) {
+            this.client = client;
+            this.mlIndicesHandler = mlIndicesHandler;
+            this.memoryManager = memoryManager;
+        }
+
+        @Override
+        public void create(Map<String, Object> map, ActionListener<ConversationIndexMemory> listener) {
+            if (map == null || map.isEmpty()) {
+                listener.onFailure(new IllegalArgumentException("Invalid input parameter for creating ConversationIndexMemory"));
+                return;
+            }
+
+            String memoryId = (String) map.get(MEMORY_ID);
+            if (Strings.isEmpty(memoryId)) {
+                listener.onFailure(new IllegalArgumentException("Memory id is required for creating ConversationIndexMemory"));
+            }
```

Comment:
If we have memory_id is not empty check here, we can go with line 201 to return memory instance directly.

Reply:
Good catch! Missed a return express, will add.